### DISCUSSION TITLE
Questions about the autofree feature

### DISCUSSION THREAD

**Discussion by RolandHeinze** [vlang/v#12343] [General]
I am new to the V Programming Language, and I read on the [homepage of V](https://vlang.io/), section "Innovative memory management":

> Most objects (~90-100%) are freed by V's autofree engine: the compiler inserts necessary free calls automatically during compilation. Remaining small percentage of objects is freed via GC.

I got some questions concering that:
1.  When exactly can't the compiler add free calls, and must use GC? Is there any specification for it, yet, or will it be added, when autofree becomes the default (V0.3)?
2. If the compiler can't add free calls and must use GC, it would be helpful to know, in which cases or source code positions the compiler was not able to add free calls but needed to use a GC. Is it possible, that the compiler outputs information, warning or error messages in such cases? Maybe it would allow to edit the source code in a way, that the compiler does not need to use a GC after the change. My target is to write 100% GC free code, and I would be happy, if the compiler assists it. Maybe the compiler output for GC usage can also be put into the V language service, and be used in the VSCode V Programming Language extension to visualize problems with autofree in VSCode.
3. How much space does the GC take in the final executable?

**Comment by qpwo** [comment]
I am very curious about this as well. See also: https://github.com/vlang/v/discussions/13588

**Comment by Wajinn** [comment]
> > Most objects (~90-100%) are freed by V's autofree engine: the compiler inserts necessary free calls automatically during compilation. Remaining small percentage of objects is freed via GC.

This has been changed to the following (https://github.com/vlang/v/blob/master/doc/docs.md).

Most objects (~90-100%) are freed by V's autofree engine: the compiler inserts necessary free calls automatically during compilation. Remaining small percentage of objects is freed via reference counting.  

That Autofree would use reference counting for that last possible ~10% looks not only more doable, but wanted as well.  Often the distinction between ARC and GC, is compile time versus runtime.  GC (which appears to be doing well so far) would be a separate option for those that want it.  Both GC and Autofree being available options, which are separate from each other, would arguably be the best case scenario for many.

**Reply by ctxcode** [reply]
In my v program, only 0.1% was autofreed and 99.9% was freed by the garbage collector. It all depends on the program you are making. The GC is still really fast though.
